{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "be3dfc14-7281-445d-b731-85703fe3b68f",
   "metadata": {},
   "source": [
    "\n",
    "In this section, we demonstrate the generation of single-cell–level regulatory graphs required for running scReGAT on unmatched multi-omics datasets.\n",
    "\n",
    "As an illustrative example, we employ human cerebral cortex data, originally reported by Mulqueen et al. (scATAC-seq) and Hodge et al. (scRNA-seq)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d94e249f-86bd-4ab0-8f7f-298838e254bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/scregat/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/opt/conda/envs/scregat/lib/python3.10/site-packages/anndata/_metadata.py:15: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import random\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import hypergeom\n",
    "from scregat.data_process import prepare_model_input, sum_counts, plot_edge, ATACGraphDataset\n",
    "import scanpy as sc\n",
    "import itertools\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a1083fe-cae0-4dcd-8c48-e76c960c1b97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/scregat/lib/python3.10/site-packages/scregat/data_process.py\n",
      "/opt/conda/envs/scregat/lib/python3.10/site-packages/scregat\n"
     ]
    }
   ],
   "source": [
    "import scregat\n",
    "import os\n",
    "\n",
    "print(scregat.data_process.__file__)    \n",
    "print(os.path.dirname(scregat.__file__))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9f6c82c5-d068-45c0-9e14-149c055ae282",
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_h5ad_file = \"../reproduce_data/HumanBrain_ATAC.h5ad\"\n",
    "RNA_h5ad_file = \"../reproduce_data/HumanBrain_RNA.h5ad\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85b5b726-50b2-4192-8cef-216e48a77b12",
   "metadata": {},
   "source": [
    "\n",
    "- It is important to note that both **adata_rna** and **adata_atac** must contain a `celltype` column in their `obs`.  \n",
    "- The cell types present in **adata_atac** must also be represented in the `obs` of **adata_rna**.  \n",
    "- The `celltype` entries in **adata_rna** are allowed to include a larger number of categories.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7efb58a5-ad59-4d17-b0fc-610cde47f240",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_atac = sc.read_h5ad(ATAC_h5ad_file)\n",
    "adata_rna = sc.read_h5ad(RNA_h5ad_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "52c9406f-3651-4272-a542-28586cf7e378",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2174 × 292156\n",
       "    obs: 'celltype', 'celltype_rna'\n",
       "    var: 'peak'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "55e70b9b-056b-41c6-8ad4-2bf5bd64c9ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>celltype_rna</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GAAGCAGCTCTGACGACCGCGGTT</th>\n",
       "      <td>inhibitory_neuron</td>\n",
       "      <td>inhibitory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCATACCAATGATGCGGCATTCT</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTGAGGAACCGGAAGACACTAAG</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGCGACGTCTGACGAGCCACAGG</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTGAGGAAGTTACGCTCCAACGC</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAAGAGTAGCCAAGGCGTGAATAT</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAAGCAGCTTAACTCAGGTACCTT</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTTAGTTAGTTACGCTTGGACTC</th>\n",
       "      <td>polydendrocytes</td>\n",
       "      <td>polydendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAAGAGTAATGGAGCTCTTGGTAT</th>\n",
       "      <td>oligodendrocytes</td>\n",
       "      <td>oligodendrocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAAGAGTAATGGAGCTATCTGCCA</th>\n",
       "      <td>microglia</td>\n",
       "      <td>microglia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2174 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   celltype       celltype_rna\n",
       "GAAGCAGCTCTGACGACCGCGGTT  inhibitory_neuron  inhibitory_neuron\n",
       "TCCATACCAATGATGCGGCATTCT   oligodendrocytes   oligodendrocytes\n",
       "ATTGAGGAACCGGAAGACACTAAG   oligodendrocytes   oligodendrocytes\n",
       "ACGCGACGTCTGACGAGCCACAGG   oligodendrocytes   oligodendrocytes\n",
       "ATTGAGGAAGTTACGCTCCAACGC   oligodendrocytes   oligodendrocytes\n",
       "...                                     ...                ...\n",
       "GAAGAGTAGCCAAGGCGTGAATAT   oligodendrocytes   oligodendrocytes\n",
       "GAAGCAGCTTAACTCAGGTACCTT   oligodendrocytes   oligodendrocytes\n",
       "GGTTAGTTAGTTACGCTTGGACTC    polydendrocytes    polydendrocytes\n",
       "GAAGAGTAATGGAGCTCTTGGTAT   oligodendrocytes   oligodendrocytes\n",
       "GAAGAGTAATGGAGCTATCTGCCA          microglia          microglia\n",
       "\n",
       "[2174 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.obs['celltype_rna'] == adata_atac.obs['celltype']\n",
    "adata_atac.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0660da80-a298-4d37-8b54-9d199d0204cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10-10113-10613</th>\n",
       "      <td>chr10:10113-10613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10-74022-74522</th>\n",
       "      <td>chr10:74022-74522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10-76384-76884</th>\n",
       "      <td>chr10:76384-76884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10-129575-130075</th>\n",
       "      <td>chr10:129575-130075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10-134348-135024</th>\n",
       "      <td>chr10:134348-135024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX-155148867-155149367</th>\n",
       "      <td>chrX:155148867-155149367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX-155215853-155217198</th>\n",
       "      <td>chrX:155215853-155217198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX-155263318-155263818</th>\n",
       "      <td>chrX:155263318-155263818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX-155263973-155264786</th>\n",
       "      <td>chrX:155263973-155264786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX-155612285-155613554</th>\n",
       "      <td>chrX:155612285-155613554</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>292156 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              peak\n",
       "chr10-10113-10613                chr10:10113-10613\n",
       "chr10-74022-74522                chr10:74022-74522\n",
       "chr10-76384-76884                chr10:76384-76884\n",
       "chr10-129575-130075            chr10:129575-130075\n",
       "chr10-134348-135024            chr10:134348-135024\n",
       "...                                            ...\n",
       "chrX-155148867-155149367  chrX:155148867-155149367\n",
       "chrX-155215853-155217198  chrX:155215853-155217198\n",
       "chrX-155263318-155263818  chrX:155263318-155263818\n",
       "chrX-155263973-155264786  chrX:155263973-155264786\n",
       "chrX-155612285-155613554  chrX:155612285-155613554\n",
       "\n",
       "[292156 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5101f4b8-29b6-4c5c-9d8d-277f7154b2e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 15603 × 50281\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'cluster', 'class', 'brain_subregion', 'donor', 'sex', 'facs_sort_criteria', 'seq_batch', 'total_reads', 'percent_reads_unique', 'celltype'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "02b4b25a-d148-4b0e-b6d7-7373207fc797",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>cluster</th>\n",
       "      <th>class</th>\n",
       "      <th>brain_subregion</th>\n",
       "      <th>donor</th>\n",
       "      <th>sex</th>\n",
       "      <th>facs_sort_criteria</th>\n",
       "      <th>seq_batch</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>percent_reads_unique</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>F1S4_160106_001_B01</th>\n",
       "      <td>0</td>\n",
       "      <td>1945567.0</td>\n",
       "      <td>8635</td>\n",
       "      <td>Inh L4-6 SST B3GAT2</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H200.1030</td>\n",
       "      <td>M</td>\n",
       "      <td>NeuN-positive</td>\n",
       "      <td>R8S4-160411-H</td>\n",
       "      <td>2572946</td>\n",
       "      <td>85.573696</td>\n",
       "      <td>inhibitory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_160106_001_C01</th>\n",
       "      <td>0</td>\n",
       "      <td>2076398.0</td>\n",
       "      <td>11697</td>\n",
       "      <td>Exc L5-6 RORB TTC12</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H200.1030</td>\n",
       "      <td>M</td>\n",
       "      <td>NeuN-positive</td>\n",
       "      <td>R8S4-160411-H</td>\n",
       "      <td>2755839</td>\n",
       "      <td>87.619777</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_160106_001_E01</th>\n",
       "      <td>0</td>\n",
       "      <td>1984845.0</td>\n",
       "      <td>12138</td>\n",
       "      <td>Exc L5-6 FEZF2 ABO</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H200.1030</td>\n",
       "      <td>M</td>\n",
       "      <td>NeuN-positive</td>\n",
       "      <td>R8S4-160411-H</td>\n",
       "      <td>2701064</td>\n",
       "      <td>86.448562</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_160106_001_G01</th>\n",
       "      <td>0</td>\n",
       "      <td>1991032.0</td>\n",
       "      <td>12191</td>\n",
       "      <td>Exc L5-6 FEZF2 EFTUD1P1</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H200.1030</td>\n",
       "      <td>M</td>\n",
       "      <td>NeuN-positive</td>\n",
       "      <td>R8S4-160411-H</td>\n",
       "      <td>2759117</td>\n",
       "      <td>84.497105</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1S4_160106_001_H01</th>\n",
       "      <td>0</td>\n",
       "      <td>2189892.0</td>\n",
       "      <td>10535</td>\n",
       "      <td>Exc L3-5 RORB ESR1</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H200.1030</td>\n",
       "      <td>M</td>\n",
       "      <td>NeuN-positive</td>\n",
       "      <td>R8S4-160411-H</td>\n",
       "      <td>2930410</td>\n",
       "      <td>87.014752</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_170405_060_B01</th>\n",
       "      <td>1</td>\n",
       "      <td>2315521.0</td>\n",
       "      <td>10853</td>\n",
       "      <td>Exc L5-6 FEZF2 ABO</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H16.06.008</td>\n",
       "      <td>F</td>\n",
       "      <td>SATB2-pos, NeuN-pos</td>\n",
       "      <td>R8S4-170505</td>\n",
       "      <td>3015621</td>\n",
       "      <td>89.829491</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_170405_060_C01</th>\n",
       "      <td>1</td>\n",
       "      <td>2376631.0</td>\n",
       "      <td>10539</td>\n",
       "      <td>Exc L5-6 THEMIS FGF10</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H16.06.008</td>\n",
       "      <td>F</td>\n",
       "      <td>SATB2-pos, NeuN-pos</td>\n",
       "      <td>R8S4-170505</td>\n",
       "      <td>3121536</td>\n",
       "      <td>89.871717</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_170405_060_E01</th>\n",
       "      <td>1</td>\n",
       "      <td>1761501.0</td>\n",
       "      <td>9017</td>\n",
       "      <td>Exc L4-6 FEZF2 IL26</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H16.06.008</td>\n",
       "      <td>F</td>\n",
       "      <td>SATB2-neg, NeuN-pos</td>\n",
       "      <td>R8S4-170505</td>\n",
       "      <td>2771628</td>\n",
       "      <td>73.988717</td>\n",
       "      <td>excitatory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_170405_060_F01</th>\n",
       "      <td>1</td>\n",
       "      <td>1736137.0</td>\n",
       "      <td>4953</td>\n",
       "      <td>Inh L4-6 SST B3GAT2</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H16.06.008</td>\n",
       "      <td>F</td>\n",
       "      <td>SATB2-neg, NeuN-pos</td>\n",
       "      <td>R8S4-170505</td>\n",
       "      <td>2307182</td>\n",
       "      <td>86.352789</td>\n",
       "      <td>inhibitory_neuron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F2S4_170405_060_H01</th>\n",
       "      <td>1</td>\n",
       "      <td>1987249.0</td>\n",
       "      <td>9978</td>\n",
       "      <td>Inh L5-6 SST MIR548F2</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>L5</td>\n",
       "      <td>H16.06.008</td>\n",
       "      <td>F</td>\n",
       "      <td>SATB2-neg, NeuN-pos</td>\n",
       "      <td>R8S4-170505</td>\n",
       "      <td>2623262</td>\n",
       "      <td>89.734803</td>\n",
       "      <td>inhibitory_neuron</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15603 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "F1S4_160106_001_B01           0   1945567.0          8635   \n",
       "F1S4_160106_001_C01           0   2076398.0         11697   \n",
       "F1S4_160106_001_E01           0   1984845.0         12138   \n",
       "F1S4_160106_001_G01           0   1991032.0         12191   \n",
       "F1S4_160106_001_H01           0   2189892.0         10535   \n",
       "...                         ...         ...           ...   \n",
       "F2S4_170405_060_B01           1   2315521.0         10853   \n",
       "F2S4_170405_060_C01           1   2376631.0         10539   \n",
       "F2S4_170405_060_E01           1   1761501.0          9017   \n",
       "F2S4_170405_060_F01           1   1736137.0          4953   \n",
       "F2S4_170405_060_H01           1   1987249.0          9978   \n",
       "\n",
       "                                     cluster          class brain_subregion  \\\n",
       "F1S4_160106_001_B01      Inh L4-6 SST B3GAT2      GABAergic              L5   \n",
       "F1S4_160106_001_C01      Exc L5-6 RORB TTC12  Glutamatergic              L5   \n",
       "F1S4_160106_001_E01       Exc L5-6 FEZF2 ABO  Glutamatergic              L5   \n",
       "F1S4_160106_001_G01  Exc L5-6 FEZF2 EFTUD1P1  Glutamatergic              L5   \n",
       "F1S4_160106_001_H01       Exc L3-5 RORB ESR1  Glutamatergic              L5   \n",
       "...                                      ...            ...             ...   \n",
       "F2S4_170405_060_B01       Exc L5-6 FEZF2 ABO  Glutamatergic              L5   \n",
       "F2S4_170405_060_C01    Exc L5-6 THEMIS FGF10  Glutamatergic              L5   \n",
       "F2S4_170405_060_E01      Exc L4-6 FEZF2 IL26  Glutamatergic              L5   \n",
       "F2S4_170405_060_F01      Inh L4-6 SST B3GAT2      GABAergic              L5   \n",
       "F2S4_170405_060_H01    Inh L5-6 SST MIR548F2      GABAergic              L5   \n",
       "\n",
       "                          donor sex   facs_sort_criteria      seq_batch  \\\n",
       "F1S4_160106_001_B01   H200.1030   M        NeuN-positive  R8S4-160411-H   \n",
       "F1S4_160106_001_C01   H200.1030   M        NeuN-positive  R8S4-160411-H   \n",
       "F1S4_160106_001_E01   H200.1030   M        NeuN-positive  R8S4-160411-H   \n",
       "F1S4_160106_001_G01   H200.1030   M        NeuN-positive  R8S4-160411-H   \n",
       "F1S4_160106_001_H01   H200.1030   M        NeuN-positive  R8S4-160411-H   \n",
       "...                         ...  ..                  ...            ...   \n",
       "F2S4_170405_060_B01  H16.06.008   F  SATB2-pos, NeuN-pos    R8S4-170505   \n",
       "F2S4_170405_060_C01  H16.06.008   F  SATB2-pos, NeuN-pos    R8S4-170505   \n",
       "F2S4_170405_060_E01  H16.06.008   F  SATB2-neg, NeuN-pos    R8S4-170505   \n",
       "F2S4_170405_060_F01  H16.06.008   F  SATB2-neg, NeuN-pos    R8S4-170505   \n",
       "F2S4_170405_060_H01  H16.06.008   F  SATB2-neg, NeuN-pos    R8S4-170505   \n",
       "\n",
       "                     total_reads  percent_reads_unique           celltype  \n",
       "F1S4_160106_001_B01      2572946             85.573696  inhibitory_neuron  \n",
       "F1S4_160106_001_C01      2755839             87.619777  excitatory_neuron  \n",
       "F1S4_160106_001_E01      2701064             86.448562  excitatory_neuron  \n",
       "F1S4_160106_001_G01      2759117             84.497105  excitatory_neuron  \n",
       "F1S4_160106_001_H01      2930410             87.014752  excitatory_neuron  \n",
       "...                          ...                   ...                ...  \n",
       "F2S4_170405_060_B01      3015621             89.829491  excitatory_neuron  \n",
       "F2S4_170405_060_C01      3121536             89.871717  excitatory_neuron  \n",
       "F2S4_170405_060_E01      2771628             73.988717  excitatory_neuron  \n",
       "F2S4_170405_060_F01      2307182             86.352789  inhibitory_neuron  \n",
       "F2S4_170405_060_H01      2623262             89.734803  inhibitory_neuron  \n",
       "\n",
       "[15603 rows x 13 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "93597ef2-fd6b-435c-b9c3-9cd24e8bc695",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3.8-1.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.8-1.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.8-1.4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.8-1.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5-HT3C2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bA255A11.4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bA395L14.12</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50281 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [3.8-1.2, 3.8-1.3, 3.8-1.4, 3.8-1.5, 5-HT3C2, A1BG, A1BG-AS1, A1CF, A2M, A2M-AS1, A2ML1, A2MP1, A3GALT2, A4GALT, A4GNT, AA06, AAAS, AACS, AACSP1, AADAC, AADACL2, AADACL2-AS1, AADACL3, AADACL4, AADACP1, AADAT, AAED1, AAGAB, AAK1, AAMDC, AAMP, AANAT, AAR2, AARD, AARS, AARS2, AARSD1, AASDH, AASDHPPT, AASS, AATBC, AATF, AATK, AATK-AS1, ABAT, ABCA1, ABCA10, ABCA11P, ABCA12, ABCA13, ABCA17P, ABCA2, ABCA3, ABCA4, ABCA5, ABCA6, ABCA7, ABCA8, ABCA9, ABCA9-AS1, ABCB1, ABCB10, ABCB10P1, ABCB10P3, ABCB10P4, ABCB11, ABCB4, ABCB5, ABCB6, ABCB7, ABCB8, ABCB9, ABCC1, ABCC10, ABCC11, ABCC12, ABCC13, ABCC2, ABCC3, ABCC4, ABCC5, ABCC5-AS1, ABCC6, ABCC6P1, ABCC6P2, ABCC8, ABCC9, ABCD1, ABCD1P1, ABCD1P2, ABCD1P3, ABCD1P4, ABCD1P5, ABCD2, ABCD3, ABCD4, ABCE1, ABCF1, ABCF2, ABCF2P1, ...]\n",
       "\n",
       "[50281 rows x 0 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ab397b57-694a-4413-af79-bc82517af277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['inhibitory_neuron', 'oligodendrocytes', 'microglia', 'excitatory_neuron', 'astrocyte', 'polydendrocytes']\n",
       "Categories (6, object): ['astrocyte', 'excitatory_neuron', 'inhibitory_neuron', 'microglia', 'oligodendrocytes', 'polydendrocytes']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.obs.celltype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aea341d9-2615-4c0c-b48e-589b489c21b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['inhibitory_neuron', 'excitatory_neuron', 'oligodendrocytes', 'OPC', 'astrocyte', 'microglia', 'polydendrocytes']\n",
       "Categories (7, object): ['OPC', 'astrocyte', 'excitatory_neuron', 'inhibitory_neuron', 'microglia', 'oligodendrocytes', 'polydendrocytes']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.obs.celltype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "19e184c6-e2fa-495c-9add-1964a6d57b27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**finished identifying marker genes by COSG**\n"
     ]
    }
   ],
   "source": [
    "adata_rna.obs['celltype'] = adata_rna.obs['celltype'].astype('object')\n",
    "df_rna = sum_counts(adata_rna,by = 'celltype', marker_gene_num=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a7e8eb66-a40a-401d-950e-26d3a291f6a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RXFP1</th>\n",
       "      <th>ATP13A4</th>\n",
       "      <th>PGR</th>\n",
       "      <th>EMC10</th>\n",
       "      <th>GALNT3</th>\n",
       "      <th>LSP1</th>\n",
       "      <th>LOC101929372</th>\n",
       "      <th>GPNMB</th>\n",
       "      <th>P2RY12</th>\n",
       "      <th>CD302</th>\n",
       "      <th>...</th>\n",
       "      <th>ATP6V0C</th>\n",
       "      <th>ABI2</th>\n",
       "      <th>TRIOBP</th>\n",
       "      <th>ICOSLG</th>\n",
       "      <th>RERG</th>\n",
       "      <th>PREX1</th>\n",
       "      <th>GSN</th>\n",
       "      <th>GAP43</th>\n",
       "      <th>SIRPB2</th>\n",
       "      <th>FAM177B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OPC</th>\n",
       "      <td>1.977282</td>\n",
       "      <td>243.189362</td>\n",
       "      <td>6.567498</td>\n",
       "      <td>49.746312</td>\n",
       "      <td>58.736322</td>\n",
       "      <td>0.230454</td>\n",
       "      <td>0.162515</td>\n",
       "      <td>324.135100</td>\n",
       "      <td>9.514092</td>\n",
       "      <td>1.739794</td>\n",
       "      <td>...</td>\n",
       "      <td>27.589659</td>\n",
       "      <td>126.436216</td>\n",
       "      <td>25.480129</td>\n",
       "      <td>0.318424</td>\n",
       "      <td>7.756298</td>\n",
       "      <td>91.241953</td>\n",
       "      <td>69.767077</td>\n",
       "      <td>30.394268</td>\n",
       "      <td>0.200216</td>\n",
       "      <td>0.057349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>astrocyte</th>\n",
       "      <td>5.460383</td>\n",
       "      <td>592.013255</td>\n",
       "      <td>1.122709</td>\n",
       "      <td>27.688592</td>\n",
       "      <td>3.063550</td>\n",
       "      <td>0.187578</td>\n",
       "      <td>30.816097</td>\n",
       "      <td>2.506366</td>\n",
       "      <td>10.546135</td>\n",
       "      <td>2.108429</td>\n",
       "      <td>...</td>\n",
       "      <td>14.400693</td>\n",
       "      <td>186.549481</td>\n",
       "      <td>30.933883</td>\n",
       "      <td>0.143088</td>\n",
       "      <td>309.865948</td>\n",
       "      <td>193.036843</td>\n",
       "      <td>180.726996</td>\n",
       "      <td>10.029325</td>\n",
       "      <td>0.033279</td>\n",
       "      <td>0.158025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>excitatory_neuron</th>\n",
       "      <td>9618.642900</td>\n",
       "      <td>525.937022</td>\n",
       "      <td>22.664902</td>\n",
       "      <td>500.026030</td>\n",
       "      <td>613.249599</td>\n",
       "      <td>48.892340</td>\n",
       "      <td>11.371342</td>\n",
       "      <td>33.073244</td>\n",
       "      <td>384.708391</td>\n",
       "      <td>0.642745</td>\n",
       "      <td>...</td>\n",
       "      <td>2896.196396</td>\n",
       "      <td>11327.085658</td>\n",
       "      <td>270.302763</td>\n",
       "      <td>2.743651</td>\n",
       "      <td>991.896530</td>\n",
       "      <td>1119.595128</td>\n",
       "      <td>978.919536</td>\n",
       "      <td>12965.963452</td>\n",
       "      <td>4.599011</td>\n",
       "      <td>33.381850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>inhibitory_neuron</th>\n",
       "      <td>293.852883</td>\n",
       "      <td>310.475710</td>\n",
       "      <td>29.598662</td>\n",
       "      <td>306.189812</td>\n",
       "      <td>106.368897</td>\n",
       "      <td>12.411594</td>\n",
       "      <td>5.664125</td>\n",
       "      <td>53.055618</td>\n",
       "      <td>57.614247</td>\n",
       "      <td>3.288799</td>\n",
       "      <td>...</td>\n",
       "      <td>2170.352186</td>\n",
       "      <td>2949.813759</td>\n",
       "      <td>148.803219</td>\n",
       "      <td>1.305105</td>\n",
       "      <td>2853.188535</td>\n",
       "      <td>186.263144</td>\n",
       "      <td>241.081847</td>\n",
       "      <td>1817.570922</td>\n",
       "      <td>4.223736</td>\n",
       "      <td>12.831230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>microglia</th>\n",
       "      <td>0.609588</td>\n",
       "      <td>5.631434</td>\n",
       "      <td>1.062473</td>\n",
       "      <td>3.883262</td>\n",
       "      <td>0.279265</td>\n",
       "      <td>0.073228</td>\n",
       "      <td>0.432567</td>\n",
       "      <td>7.109798</td>\n",
       "      <td>137.660817</td>\n",
       "      <td>2.706166</td>\n",
       "      <td>...</td>\n",
       "      <td>10.098409</td>\n",
       "      <td>9.971359</td>\n",
       "      <td>8.799556</td>\n",
       "      <td>0.590807</td>\n",
       "      <td>0.054346</td>\n",
       "      <td>78.407873</td>\n",
       "      <td>45.765370</td>\n",
       "      <td>0.680299</td>\n",
       "      <td>16.595570</td>\n",
       "      <td>49.648214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oligodendrocytes</th>\n",
       "      <td>8.839663</td>\n",
       "      <td>36.150258</td>\n",
       "      <td>3.003144</td>\n",
       "      <td>161.119213</td>\n",
       "      <td>1.386353</td>\n",
       "      <td>1.194011</td>\n",
       "      <td>2.254029</td>\n",
       "      <td>2.276863</td>\n",
       "      <td>1.573197</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>57.737471</td>\n",
       "      <td>124.412413</td>\n",
       "      <td>23.632928</td>\n",
       "      <td>7.546591</td>\n",
       "      <td>3.224950</td>\n",
       "      <td>303.360678</td>\n",
       "      <td>427.687454</td>\n",
       "      <td>18.605865</td>\n",
       "      <td>0.832426</td>\n",
       "      <td>3.697203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>polydendrocytes</th>\n",
       "      <td>0.022247</td>\n",
       "      <td>2.643582</td>\n",
       "      <td>5.326460</td>\n",
       "      <td>0.011465</td>\n",
       "      <td>0.006362</td>\n",
       "      <td>1.855250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031521</td>\n",
       "      <td>0.011087</td>\n",
       "      <td>0.596851</td>\n",
       "      <td>...</td>\n",
       "      <td>1.539264</td>\n",
       "      <td>3.578376</td>\n",
       "      <td>6.383234</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.436578</td>\n",
       "      <td>8.039381</td>\n",
       "      <td>7.974619</td>\n",
       "      <td>1.846928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 2031 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         RXFP1     ATP13A4        PGR       EMC10      GALNT3  \\\n",
       "celltype                                                                        \n",
       "OPC                   1.977282  243.189362   6.567498   49.746312   58.736322   \n",
       "astrocyte             5.460383  592.013255   1.122709   27.688592    3.063550   \n",
       "excitatory_neuron  9618.642900  525.937022  22.664902  500.026030  613.249599   \n",
       "inhibitory_neuron   293.852883  310.475710  29.598662  306.189812  106.368897   \n",
       "microglia             0.609588    5.631434   1.062473    3.883262    0.279265   \n",
       "oligodendrocytes      8.839663   36.150258   3.003144  161.119213    1.386353   \n",
       "polydendrocytes       0.022247    2.643582   5.326460    0.011465    0.006362   \n",
       "\n",
       "                        LSP1  LOC101929372       GPNMB      P2RY12     CD302  \\\n",
       "celltype                                                                       \n",
       "OPC                 0.230454      0.162515  324.135100    9.514092  1.739794   \n",
       "astrocyte           0.187578     30.816097    2.506366   10.546135  2.108429   \n",
       "excitatory_neuron  48.892340     11.371342   33.073244  384.708391  0.642745   \n",
       "inhibitory_neuron  12.411594      5.664125   53.055618   57.614247  3.288799   \n",
       "microglia           0.073228      0.432567    7.109798  137.660817  2.706166   \n",
       "oligodendrocytes    1.194011      2.254029    2.276863    1.573197  0.000000   \n",
       "polydendrocytes     1.855250      0.000000    0.031521    0.011087  0.596851   \n",
       "\n",
       "                   ...      ATP6V0C          ABI2      TRIOBP    ICOSLG  \\\n",
       "celltype           ...                                                    \n",
       "OPC                ...    27.589659    126.436216   25.480129  0.318424   \n",
       "astrocyte          ...    14.400693    186.549481   30.933883  0.143088   \n",
       "excitatory_neuron  ...  2896.196396  11327.085658  270.302763  2.743651   \n",
       "inhibitory_neuron  ...  2170.352186   2949.813759  148.803219  1.305105   \n",
       "microglia          ...    10.098409      9.971359    8.799556  0.590807   \n",
       "oligodendrocytes   ...    57.737471    124.412413   23.632928  7.546591   \n",
       "polydendrocytes    ...     1.539264      3.578376    6.383234  0.000000   \n",
       "\n",
       "                          RERG        PREX1         GSN         GAP43  \\\n",
       "celltype                                                                \n",
       "OPC                   7.756298    91.241953   69.767077     30.394268   \n",
       "astrocyte           309.865948   193.036843  180.726996     10.029325   \n",
       "excitatory_neuron   991.896530  1119.595128  978.919536  12965.963452   \n",
       "inhibitory_neuron  2853.188535   186.263144  241.081847   1817.570922   \n",
       "microglia             0.054346    78.407873   45.765370      0.680299   \n",
       "oligodendrocytes      3.224950   303.360678  427.687454     18.605865   \n",
       "polydendrocytes       6.436578     8.039381    7.974619      1.846928   \n",
       "\n",
       "                      SIRPB2    FAM177B  \n",
       "celltype                                 \n",
       "OPC                 0.200216   0.057349  \n",
       "astrocyte           0.033279   0.158025  \n",
       "excitatory_neuron   4.599011  33.381850  \n",
       "inhibitory_neuron   4.223736  12.831230  \n",
       "microglia          16.595570  49.648214  \n",
       "oligodendrocytes    0.832426   3.697203  \n",
       "polydendrocytes     0.000000   0.000000  \n",
       "\n",
       "[7 rows x 2031 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b940e46a-f6dd-4163-9eff-c3991e00d762",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "28f87e13-c036-481b-9038-84964660aa59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PO.txt',\n",
       " 'all_tissue_SNP_Gene.txt',\n",
       " 'TF_Gene_tissue_cutoff1.csv',\n",
       " 'PO_brain.txt',\n",
       " 'genes.protein.tss.tsv',\n",
       " 'hg38.chrom.sizes',\n",
       " 'PP.txt',\n",
       " 'readme.md',\n",
       " 'trrust_rawdata.human.tsv',\n",
       " 'celltype_specific_cRE_interactions',\n",
       " 'PP_brain.txt',\n",
       " 'TF_Gene_tissue_Brain.csv',\n",
       " 'model_init.pth',\n",
       " 'dataset_atac_kRG_Pancreas.pkl',\n",
       " 'processed_files',\n",
       " 'peaks_process']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scregat\n",
    "import os\n",
    "# This step adds tissue-specific Hi-C regulatory relationships.\n",
    "# The user needs to provide a set of files. For example,\n",
    "#in \n",
    "base_dir = '../data/'\n",
    "os.listdir(base_dir)\n",
    "# we include a file called PO_brain.txt,\n",
    "# which contains brain tissue-specific Hi-C links.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9f998ce1-4dea-4b29-a0dc-863b639716a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "only dataset_obj ...\n",
      "processing Hi-C ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " SettingWithCopyWarning:/opt/conda/envs/scregat/lib/python3.10/site-packages/scregat/data_process.py:607: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing TF ...\n",
      "additional TF...\n",
      "total candidate tf-gene:  28054\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/opt/conda/envs/scregat/lib/python3.10/site-packages/scregat/data_process.py:712: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:274.)\n"
     ]
    }
   ],
   "source": [
    "dataset_obj = prepare_model_input(\n",
    "    # [Core Data] Single-cell ATAC-seq AnnData object.\n",
    "    # Requirement: .X must be Peak-by-Cell matrix; .obs must contain cell type annotations.\n",
    "    adata_atac = adata_atac,\n",
    "    \n",
    "    # [Output Path] Root directory for storing intermediate processed files.\n",
    "    # The script will create a 'processed_files' folder here (e.g., sorted bed files).\n",
    "    path_data_root = '../',\n",
    "    \n",
    "    # [File Reference] String path to the original ATAC file (used for naming/logging).\n",
    "    file_atac = ATAC_h5ad_file, \n",
    "    \n",
    "    # [Core Data] RNA expression matrix aggregated by cell type (Pseudo-bulk).\n",
    "    # Format: Pandas DataFrame. \n",
    "    # Index (Rows): Cell type names (must match 'celltype_rna' in adata_atac.obs).\n",
    "    # Columns: Gene Symbols (e.g., 'TP53').\n",
    "    df_rna_celltype = df_rna,\n",
    "    \n",
    "    # [Prior Knowledge] Path to eQTL (expression Quantitative Trait Loci) data.\n",
    "    # Used to link SNPs/non-coding regions to target genes.\n",
    "    path_eqtl = '../data/all_tissue_SNP_Gene.txt',\n",
    "    \n",
    "    # [Prior Knowledge] Suffix for Hi-C interaction files to specify tissue context.\n",
    "    # e.g., if set to \"_brain\", the code looks for 'PP_brain.txt' and 'PO_brain.txt'.\n",
    "    # PP = Promoter-Promoter, PO = Promoter-Other (Enhancer).\n",
    "    Hi_C_file_suffix = \"_\" + \"brain\",  \n",
    "    \n",
    "    # [Preprocessing] Whether to convert genomic coordinates from hg19 to hg38.\n",
    "    # Set True if input ATAC peaks are hg19 (requires LiftOver tool); False if already hg38.\n",
    "    hg19tohg38 = False,\n",
    "    data_dir = '../data',\n",
    "    # [QC Filter] Peak filtering threshold.\n",
    "    # Peaks must be accessible in at least 1% (0.01) of cells to be retained.\n",
    "    min_percent = 0.01,\n",
    "    \n",
    "    # [Prior Knowledge] Whether to use an extended TF (Transcription Factor) database.\n",
    "    # False: Uses TRRUST only (Curated, high-confidence).\n",
    "    # True: Uses TRRUST + CHEA3/ChIP-seq aggregated data.\n",
    "    use_additional_tf = True,\n",
    "    \n",
    "    # [Prior Knowledge] Reliability threshold for the extended TF database.\n",
    "    # Only applies if use_additional_tf=True.\n",
    "    # Integer indicating in how many tissues/datasets the TF-Gene link must appear to be kept.\n",
    "    # 10 indicates a high-confidence, conserved regulatory relationship.\n",
    "    tissue_cuttof = 10\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "148a168c-82b0-44ae-8e3f-a06a24f5d5de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Data(x=[61528, 1], edge_index=[2, 69180], y=[1], edge_tf=[155, 2], y_exp=[1348], cell='TCCATACCAATGATGCGGCATTCT')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_obj.list_graph[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3bfb0e25-d19a-455a-b7c6-45e5a3d5bf95",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_atac_test = os.path.join('./', 'dataset_atac_kRG_MFG.pkl')\n",
    "with open(file_atac_test, 'wb') as w_pkl:\n",
    "    str_pkl = pickle.dumps(dataset_obj)\n",
    "    w_pkl.write(str_pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea2e89a5-b2b7-42a4-b784-9664b4ff7cf4",
   "metadata": {},
   "source": [
    "### Add Tissue-specific TF-gene  (from CHEA3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "98d24e77-ce4e-4b02-a5b8-d1b703311364",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"../data/TF_Gene_tissue_Brain.csv\", index_col=0)\n",
    "df.columns = ['TF', 'TargetGene', 'tissue_count']\n",
    "df_tf = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "795ebcb6-285b-403c-9862-c72b85e20d20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>TargetGene</th>\n",
       "      <th>tissue_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>KDM4B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>PLXNA1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>LYPD6B</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>ESRP1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>MAB21L1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF TargetGene  tissue_count\n",
       "0  TFAP2A      KDM4B             1\n",
       "1  TFAP2A     PLXNA1             1\n",
       "2  TFAP2A     LYPD6B             1\n",
       "3  TFAP2A      ESRP1             1\n",
       "4  TFAP2A    MAB21L1             1"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6d2c31a6-7492-45dd-9146-5330c7b98a8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract all gene names from the RNA dataframe columns\n",
    "gene_list = list(dataset_obj.df_rna.columns)\n",
    "\n",
    "# Convert gene list to a set for faster membership checks\n",
    "set_gene = set(gene_list)\n",
    "\n",
    "# Filter TF dataframe to keep only rows where both TF and TargetGene are in the gene set\n",
    "tf_base_filtered = df_tf[df_tf['TF'].isin(set_gene) & df_tf['TargetGene'].isin(set_gene)]\n",
    "\n",
    "# Generate all possible gene pairs (Cartesian product of gene set with itself)\n",
    "connections = [pair for pair in itertools.product(set_gene, set_gene)]\n",
    "gene_pair_base = connections\n",
    "\n",
    "# Collect unique TFs and TargetGenes from the filtered dataframe\n",
    "tf_map_gene = set(tf_base_filtered['TF'].unique())\n",
    "target_map_gene = set(tf_base_filtered['TargetGene'].unique())\n",
    "\n",
    "# Create tuples of (TF, TargetGene) from the filtered dataframe\n",
    "tf_base_tuples = set(zip(tf_base_filtered['TF'], tf_base_filtered['TargetGene']))\n",
    "\n",
    "# Intersect with all possible gene pairs to keep only valid TF-target pairs\n",
    "map_pair = tf_base_tuples.intersection(gene_pair_base)\n",
    "\n",
    "# Convert the set of valid pairs into a list\n",
    "map_pair_list = list(map_pair)\n",
    "\n",
    "# Create a new dataframe from the valid TF-target pairs\n",
    "df_tf_new = pd.DataFrame(map_pair_list, columns=['TF', 'TargetGene'])\n",
    "\n",
    "# Concatenate the new TF-target dataframe with the existing one in dataset_obj\n",
    "df_tf_all = pd.concat([df_tf_new, dataset_obj.df_tf])\n",
    "\n",
    "# Remove duplicate rows to ensure uniqueness\n",
    "df_tf_all = df_tf_all.drop_duplicates()\n",
    "\n",
    "# Update dataset_obj with the merged TF-target dataframe\n",
    "dataset_obj.df_tf = df_tf_all\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "51484af1-fc18-4153-9c59-703c9b84e054",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dictionary to store the index of each element in dataset_atac.array_peak\n",
    "peak_index_dict = {peak: idx for idx, peak in enumerate(dataset_obj.array_peak)}\n",
    "\n",
    "# Initialize lists to store indices\n",
    "index_1 = []\n",
    "index_2 = []\n",
    "\n",
    "# Iterate over 'TF' and 'TargetGene' columns in dataset_atac.df_tf\n",
    "for k1, k2 in zip(dataset_obj.df_tf['TF'].values, dataset_obj.df_tf['TargetGene'].values):\n",
    "    # Use the dictionary to quickly retrieve indices\n",
    "    index_1.append(peak_index_dict[k1])\n",
    "    index_2.append(peak_index_dict[k2])\n",
    "\n",
    "# Stack the two index lists column-wise and convert to a PyTorch tensor\n",
    "tf_edge_vec = torch.tensor(np.vstack([index_1, index_2]).T)\n",
    "\n",
    "# Assign the TF edge tensor to the edge_tf attribute of each graph in the list\n",
    "for t in dataset_obj.list_graph:\n",
    "    t.edge_tf = tf_edge_vec\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "023a5616-3021-4c9f-a000-fb1c31ab1056",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Data(x=[61528, 1], edge_index=[2, 69180], y=[1], edge_tf=[4036, 2], y_exp=[1348], cell='TCCATACCAATGATGCGGCATTCT')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_obj.list_graph[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a4e74390-7db9-4ccf-9df2-a81cd3f30583",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_atac_test = os.path.join('./', 'dataset_atac_kRG_MFG.pkl')\n",
    "with open(file_atac_test, 'wb') as w_pkl:\n",
    "    str_pkl = pickle.dumps(dataset_obj)\n",
    "    w_pkl.write(str_pkl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb03058e-9092-4e07-ae64-4c8ed825de69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4cddf54d-8d88-4a5d-8ddd-0a32209b362b",
   "metadata": {},
   "source": [
    "\n",
    "## Training the Model and Obtaining Regulatory Scores\n",
    "\n",
    "You can train the model and generate regulatory scores directly from the command line.  \n",
    "For example, we used the following command to produce a sample dataset and a model weight file:\n",
    "\n",
    "```bash\n",
    "python ./scReGAT/run_scregat_cli.py \\\n",
    "  --input_file ./dataset_atac_kRG_MFG.pkl \\\n",
    "  --output_file ./RS_score.h5ad \\\n",
    "  --save_model_path ./scReGAT/data/model_init.pth \\\n",
    "  --gpu 2\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "## Argument Description\n",
    "\n",
    "```python\n",
    "def parse_args():\n",
    "    parser = argparse.ArgumentParser(description=\"Run scReGAT Model Training and Inference\")\n",
    "\n",
    "    # --- I/O parameters ---\n",
    "    parser.add_argument('--input_file', type=str, required=True, \n",
    "                        help='Path to the input ATAC pickle file (e.g., dataset_atac_core_MFG.pkl)')\n",
    "    parser.add_argument('--output_file', type=str, required=True, \n",
    "                        help='Path to save the output AnnData file (e.g., result.h5ad)')\n",
    "\n",
    "    # --- Single-cell expression integration parameters ---\n",
    "    parser.add_argument('--use_sc_exp', action='store_true',\n",
    "                        help='Enable integration of single-cell RNA expression data into graph node features')\n",
    "    parser.add_argument('--rna_file', type=str, default=None,\n",
    "                        help='Path to the RNA .h5ad file; required if --use_sc_exp is set')\n",
    "\n",
    "    # --- Model saving and loading ---\n",
    "    parser.add_argument('--save_model_path', type=str, default=None, \n",
    "                        help='Optional path to save trained model parameters')\n",
    "    parser.add_argument('--load_model_path', type=str, default=None, \n",
    "                        help='Optional path to load pre-trained model parameters')\n",
    "\n",
    "    # --- Training control ---\n",
    "    parser.add_argument('--skip_train', action='store_true',\n",
    "                        help='Skip the training phase and run inference directly')\n",
    "    parser.add_argument('--seed', type=int, default=1233, \n",
    "                        help='Random seed (default: 1233)')\n",
    "    parser.add_argument('--epochs', type=int, default=4, \n",
    "                        help='Number of training epochs (default: 4)')\n",
    "    parser.add_argument('--lr', type=float, default=1e-4, \n",
    "                        help='Learning rate (default: 1e-4)')\n",
    "    parser.add_argument('--batch_size', type=int, default=15, \n",
    "                        help='Training batch size (default: 15)')\n",
    "    parser.add_argument('--sparse_loss_weight', type=float, default=0.1, \n",
    "                        help='Weight for sparse loss (default: 0.1)')\n",
    "\n",
    "    # --- Testing / inference parameters ---\n",
    "    parser.add_argument('--test_batch_size', type=int, default=20, \n",
    "                        help='Batch size for inference (default: 20)')\n",
    "    parser.add_argument('--test_ratio', type=float, default=0.5, \n",
    "                        help='Ratio of cells to use for testing (default: 0.5)')\n",
    "\n",
    "    # --- Hardware parameters ---\n",
    "    parser.add_argument('--gpu', type=int, default=1, \n",
    "                        help='GPU ID to use; set -1 to use CPU (default: 1)')\n",
    "\n",
    "    return parser.parse_args()\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "## Recommendation\n",
    "\n",
    "We recommend users initialize training on new datasets using the model weight file:\n",
    "\n",
    "```\n",
    "./scReGAT/data/model_init.pth\n",
    "```\n",
    "\n",
    "This provides a stable starting point for training and helps ensure stability across experiments.\n",
    "\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7b5b215",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scregat",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
