{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "be3dfc14-7281-445d-b731-85703fe3b68f",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "In this section, we demonstrate the generation of single-cell–level regulatory graphs required for running scReGAT on matched multi-omics datasets.\n",
    "\n",
    "As an illustrative example, we employ human pancreatic tissue data, originally reported by Wang et al.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d94e249f-86bd-4ab0-8f7f-298838e254bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/scReGAT/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/opt/conda/envs/scReGAT/lib/python3.10/site-packages/anndata/_metadata.py:15: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import random\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import hypergeom\n",
    "from scregat.data_process import prepare_model_input, sum_counts, plot_edge, ATACGraphDataset\n",
    "import scanpy as sc\n",
    "import itertools\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2a1083fe-cae0-4dcd-8c48-e76c960c1b97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/root/scReGAT/scregat/data_process.py\n",
      "/root/scReGAT/scregat\n"
     ]
    }
   ],
   "source": [
    "import scregat\n",
    "import os\n",
    "\n",
    "print(scregat.data_process.__file__)\n",
    "print(os.path.dirname(scregat.__file__))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9f6c82c5-d068-45c0-9e14-149c055ae282",
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC_h5ad_file = \"../../C0027_atac_celltype.h5ad\"\n",
    "RNA_h5ad_file = \"../../C0027_rna_celltype.h5ad\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85b5b726-50b2-4192-8cef-216e48a77b12",
   "metadata": {},
   "source": [
    "\n",
    "# - It is important to note that both **adata_rna** and **adata_atac** must contain a `celltype` column in their `obs`.  \n",
    "# - The cell types present in **adata_atac** must also be represented in the `obs` of **adata_rna**.  \n",
    "# - The `celltype` entries in **adata_rna** are allowed to include a larger number of categories.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7efb58a5-ad59-4d17-b0fc-610cde47f240",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/opt/conda/envs/scReGAT/lib/python3.10/site-packages/anndata/_core/anndata.py:1840: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "adata_atac = sc.read_h5ad(ATAC_h5ad_file)\n",
    "adata_rna = sc.read_h5ad(RNA_h5ad_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52c9406f-3651-4272-a542-28586cf7e378",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6156 × 89233\n",
       "    obs: 'donor', 'celltype', 'celltype_rna'\n",
       "    var: '0', '1', '2', '3', '4', '5'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "55e70b9b-056b-41c6-8ad4-2bf5bd64c9ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "      <th>celltype</th>\n",
       "      <th>celltype_rna</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C0027_AAACAGCCACACTAAT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACAGCCATCATGGC-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACAGCCATGAGTTT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Alpha</td>\n",
       "      <td>Alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACATGCACTGGCTG-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACATGCAGGACACA-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTGCTCCGT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTGTGTCCC-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTTAGTGAT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Acinar</td>\n",
       "      <td>Acinar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTTTCGCGC-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTTTGACCT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6156 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          donor celltype celltype_rna\n",
       "C0027_AAACAGCCACACTAAT-1  C0027     Beta         Beta\n",
       "C0027_AAACAGCCATCATGGC-1  C0027     Beta         Beta\n",
       "C0027_AAACAGCCATGAGTTT-1  C0027    Alpha        Alpha\n",
       "C0027_AAACATGCACTGGCTG-1  C0027     Beta         Beta\n",
       "C0027_AAACATGCAGGACACA-1  C0027     Beta         Beta\n",
       "...                         ...      ...          ...\n",
       "C0027_TTTGTTGGTGCTCCGT-1  C0027     Beta         Beta\n",
       "C0027_TTTGTTGGTGTGTCCC-1  C0027     Beta         Beta\n",
       "C0027_TTTGTTGGTTAGTGAT-1  C0027   Acinar       Acinar\n",
       "C0027_TTTGTTGGTTTCGCGC-1  C0027     Beta         Beta\n",
       "C0027_TTTGTTGGTTTGACCT-1  C0027     Beta         Beta\n",
       "\n",
       "[6156 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.obs['celltype_rna'] == adata_atac.obs['celltype']\n",
    "adata_atac.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0660da80-a298-4d37-8b54-9d199d0204cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1-9816-10734</th>\n",
       "      <td>chr1:9816-10734</td>\n",
       "      <td>chr1:9816-10734</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9816</td>\n",
       "      <td>10734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-14483-15389</th>\n",
       "      <td>chr1:14483-15389</td>\n",
       "      <td>chr1:14483-15389</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14483</td>\n",
       "      <td>15389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-17065-17950</th>\n",
       "      <td>chr1:17065-17950</td>\n",
       "      <td>chr1:17065-17950</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17065</td>\n",
       "      <td>17950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-28812-29619</th>\n",
       "      <td>chr1:28812-29619</td>\n",
       "      <td>chr1:28812-29619</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28812</td>\n",
       "      <td>29619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-235307-236151</th>\n",
       "      <td>chr1:235307-236151</td>\n",
       "      <td>chr1:235307-236151</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>chr1</td>\n",
       "      <td>235307</td>\n",
       "      <td>236151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000192.1-510831-511698</th>\n",
       "      <td>GL000192.1:510831-511698</td>\n",
       "      <td>GL000192.1:510831-511698</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GL000192.1</td>\n",
       "      <td>510831</td>\n",
       "      <td>511698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000192.1-512746-513650</th>\n",
       "      <td>GL000192.1:512746-513650</td>\n",
       "      <td>GL000192.1:512746-513650</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GL000192.1</td>\n",
       "      <td>512746</td>\n",
       "      <td>513650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000192.1-538992-539887</th>\n",
       "      <td>GL000192.1:538992-539887</td>\n",
       "      <td>GL000192.1:538992-539887</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GL000192.1</td>\n",
       "      <td>538992</td>\n",
       "      <td>539887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000192.1-540481-541352</th>\n",
       "      <td>GL000192.1:540481-541352</td>\n",
       "      <td>GL000192.1:540481-541352</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GL000192.1</td>\n",
       "      <td>540481</td>\n",
       "      <td>541352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000192.1-544880-545752</th>\n",
       "      <td>GL000192.1:544880-545752</td>\n",
       "      <td>GL000192.1:544880-545752</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GL000192.1</td>\n",
       "      <td>544880</td>\n",
       "      <td>545752</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89233 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 0                         1  \\\n",
       "chr1-9816-10734                    chr1:9816-10734           chr1:9816-10734   \n",
       "chr1-14483-15389                  chr1:14483-15389          chr1:14483-15389   \n",
       "chr1-17065-17950                  chr1:17065-17950          chr1:17065-17950   \n",
       "chr1-28812-29619                  chr1:28812-29619          chr1:28812-29619   \n",
       "chr1-235307-236151              chr1:235307-236151        chr1:235307-236151   \n",
       "...                                            ...                       ...   \n",
       "GL000192.1-510831-511698  GL000192.1:510831-511698  GL000192.1:510831-511698   \n",
       "GL000192.1-512746-513650  GL000192.1:512746-513650  GL000192.1:512746-513650   \n",
       "GL000192.1-538992-539887  GL000192.1:538992-539887  GL000192.1:538992-539887   \n",
       "GL000192.1-540481-541352  GL000192.1:540481-541352  GL000192.1:540481-541352   \n",
       "GL000192.1-544880-545752  GL000192.1:544880-545752  GL000192.1:544880-545752   \n",
       "\n",
       "                              2           3       4       5  \n",
       "chr1-9816-10734           Peaks        chr1    9816   10734  \n",
       "chr1-14483-15389          Peaks        chr1   14483   15389  \n",
       "chr1-17065-17950          Peaks        chr1   17065   17950  \n",
       "chr1-28812-29619          Peaks        chr1   28812   29619  \n",
       "chr1-235307-236151        Peaks        chr1  235307  236151  \n",
       "...                         ...         ...     ...     ...  \n",
       "GL000192.1-510831-511698  Peaks  GL000192.1  510831  511698  \n",
       "GL000192.1-512746-513650  Peaks  GL000192.1  512746  513650  \n",
       "GL000192.1-538992-539887  Peaks  GL000192.1  538992  539887  \n",
       "GL000192.1-540481-541352  Peaks  GL000192.1  540481  541352  \n",
       "GL000192.1-544880-545752  Peaks  GL000192.1  544880  545752  \n",
       "\n",
       "[89233 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5101f4b8-29b6-4c5c-9d8d-277f7154b2e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6156 × 60658\n",
       "    obs: 'donor', 'celltype', 'celltype_rna'\n",
       "    var: '0', '1', '2', '3', '4', '5'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "02b4b25a-d148-4b0e-b6d7-7373207fc797",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "      <th>celltype</th>\n",
       "      <th>celltype_rna</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C0027_AAACAGCCACACTAAT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACAGCCATCATGGC-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACAGCCATGAGTTT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Alpha</td>\n",
       "      <td>Alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACATGCACTGGCTG-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_AAACATGCAGGACACA-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTGCTCCGT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTGTGTCCC-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTTAGTGAT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Acinar</td>\n",
       "      <td>Acinar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTTTCGCGC-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C0027_TTTGTTGGTTTGACCT-1</th>\n",
       "      <td>C0027</td>\n",
       "      <td>Beta</td>\n",
       "      <td>Beta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6156 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          donor celltype celltype_rna\n",
       "C0027_AAACAGCCACACTAAT-1  C0027     Beta         Beta\n",
       "C0027_AAACAGCCATCATGGC-1  C0027     Beta         Beta\n",
       "C0027_AAACAGCCATGAGTTT-1  C0027    Alpha        Alpha\n",
       "C0027_AAACATGCACTGGCTG-1  C0027     Beta         Beta\n",
       "C0027_AAACATGCAGGACACA-1  C0027     Beta         Beta\n",
       "...                         ...      ...          ...\n",
       "C0027_TTTGTTGGTGCTCCGT-1  C0027     Beta         Beta\n",
       "C0027_TTTGTTGGTGTGTCCC-1  C0027     Beta         Beta\n",
       "C0027_TTTGTTGGTTAGTGAT-1  C0027   Acinar       Acinar\n",
       "C0027_TTTGTTGGTTTCGCGC-1  C0027     Beta         Beta\n",
       "C0027_TTTGTTGGTTTGACCT-1  C0027     Beta         Beta\n",
       "\n",
       "[6156 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "93597ef2-fd6b-435c-b9c3-9cd24e8bc695",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>ENSG00000223972.5_2</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>12010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>ENSG00000227232.5_2</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29569</td>\n",
       "      <td>29570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.3</th>\n",
       "      <td>ENSG00000243485.5_4</td>\n",
       "      <td>RP11-34P13.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29553</td>\n",
       "      <td>30267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613.2_2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>36080</td>\n",
       "      <td>36081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>ENSG00000268020.3_4</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52472</td>\n",
       "      <td>52473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH507-145C22.3</th>\n",
       "      <td>ENSG00000278878.1_4</td>\n",
       "      <td>CH507-145C22.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>137994</td>\n",
       "      <td>137995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH17-351M24.1</th>\n",
       "      <td>ENSG00000263278.2_4</td>\n",
       "      <td>CH17-351M24.1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GL000204.1</td>\n",
       "      <td>49838</td>\n",
       "      <td>49839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH507-513H4.4</th>\n",
       "      <td>ENSG00000280614.1_4</td>\n",
       "      <td>CH507-513H4.4</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GL000220.1</td>\n",
       "      <td>154724</td>\n",
       "      <td>154725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CH507-513H4.3</th>\n",
       "      <td>ENSG00000281181.1_4</td>\n",
       "      <td>CH507-513H4.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GL000220.1</td>\n",
       "      <td>154724</td>\n",
       "      <td>154725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-555K2.3</th>\n",
       "      <td>ENSG00000280081.3_4</td>\n",
       "      <td>RP11-555K2.3</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GL000193.1</td>\n",
       "      <td>77036</td>\n",
       "      <td>81323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60658 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  0               1                2  \\\n",
       "DDX11L1         ENSG00000223972.5_2         DDX11L1  Gene Expression   \n",
       "WASH7P          ENSG00000227232.5_2          WASH7P  Gene Expression   \n",
       "RP11-34P13.3    ENSG00000243485.5_4    RP11-34P13.3  Gene Expression   \n",
       "FAM138A         ENSG00000237613.2_2         FAM138A  Gene Expression   \n",
       "OR4G4P          ENSG00000268020.3_4          OR4G4P  Gene Expression   \n",
       "...                             ...             ...              ...   \n",
       "CH507-145C22.3  ENSG00000278878.1_4  CH507-145C22.3  Gene Expression   \n",
       "CH17-351M24.1   ENSG00000263278.2_4   CH17-351M24.1  Gene Expression   \n",
       "CH507-513H4.4   ENSG00000280614.1_4   CH507-513H4.4  Gene Expression   \n",
       "CH507-513H4.3   ENSG00000281181.1_4   CH507-513H4.3  Gene Expression   \n",
       "RP11-555K2.3    ENSG00000280081.3_4    RP11-555K2.3  Gene Expression   \n",
       "\n",
       "                         3       4       5  \n",
       "DDX11L1               chr1   11868   12010  \n",
       "WASH7P                chr1   29569   29570  \n",
       "RP11-34P13.3          chr1   29553   30267  \n",
       "FAM138A               chr1   36080   36081  \n",
       "OR4G4P                chr1   52472   52473  \n",
       "...                    ...     ...     ...  \n",
       "CH507-145C22.3  GL000195.1  137994  137995  \n",
       "CH17-351M24.1   GL000204.1   49838   49839  \n",
       "CH507-513H4.4   GL000220.1  154724  154725  \n",
       "CH507-513H4.3   GL000220.1  154724  154725  \n",
       "RP11-555K2.3    GL000193.1   77036   81323  \n",
       "\n",
       "[60658 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ab397b57-694a-4413-af79-bc82517af277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Beta', 'Alpha', 'Delta', 'Acinar', 'Immune', 'Stellate', 'Ductal', 'Gamma']\n",
       "Categories (8, object): ['Acinar', 'Alpha', 'Beta', 'Delta', 'Ductal', 'Gamma', 'Immune', 'Stellate']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac.obs.celltype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "aea341d9-2615-4c0c-b48e-589b489c21b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Beta', 'Alpha', 'Delta', 'Acinar', 'Immune', 'Stellate', 'Ductal', 'Gamma']\n",
       "Categories (8, object): ['Acinar', 'Alpha', 'Beta', 'Delta', 'Ductal', 'Gamma', 'Immune', 'Stellate']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.obs.celltype.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0db7654f-7efd-4a2a-91a2-b5963ae8977c",
   "metadata": {},
   "source": [
    "# Check whether the cell type order matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "54df8a8f-af6a-451e-a267-32db5235eac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, ...,  True,  True,  True])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna.obs_names == adata_atac.obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "35a31e8c-f538-430b-961c-6f1335ef738b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/opt/conda/envs/scReGAT/lib/python3.10/site-packages/anndata/utils.py:220: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-5', 'SNORD116-6']\n"
     ]
    }
   ],
   "source": [
    "adata_rna.var_names_make_unique()\n",
    "adata_atac.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "19e184c6-e2fa-495c-9add-1964a6d57b27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**finished identifying marker genes by COSG**\n"
     ]
    }
   ],
   "source": [
    "adata_rna.obs['celltype'] = adata_rna.obs['celltype'].astype('object')\n",
    "df_rna = sum_counts(adata_rna,by = 'celltype', marker_gene_num=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a7e8eb66-a40a-401d-950e-26d3a291f6a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RYR2</th>\n",
       "      <th>DMD</th>\n",
       "      <th>AC068491.1</th>\n",
       "      <th>PDGFD</th>\n",
       "      <th>LMO4</th>\n",
       "      <th>CLCN5</th>\n",
       "      <th>LYPLAL1</th>\n",
       "      <th>C12orf79</th>\n",
       "      <th>MT-ND1</th>\n",
       "      <th>ZRANB2</th>\n",
       "      <th>...</th>\n",
       "      <th>CLTC</th>\n",
       "      <th>RASGRF2</th>\n",
       "      <th>PCLO</th>\n",
       "      <th>CPA1</th>\n",
       "      <th>DEPTOR</th>\n",
       "      <th>MYO10</th>\n",
       "      <th>TTC32</th>\n",
       "      <th>ME1</th>\n",
       "      <th>KCTD16</th>\n",
       "      <th>ETV1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Acinar</th>\n",
       "      <td>98</td>\n",
       "      <td>239</td>\n",
       "      <td>30</td>\n",
       "      <td>322</td>\n",
       "      <td>136</td>\n",
       "      <td>73</td>\n",
       "      <td>348</td>\n",
       "      <td>1244</td>\n",
       "      <td>159</td>\n",
       "      <td>246</td>\n",
       "      <td>...</td>\n",
       "      <td>229</td>\n",
       "      <td>5</td>\n",
       "      <td>270</td>\n",
       "      <td>408</td>\n",
       "      <td>55</td>\n",
       "      <td>883</td>\n",
       "      <td>18</td>\n",
       "      <td>126</td>\n",
       "      <td>132</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alpha</th>\n",
       "      <td>125</td>\n",
       "      <td>10725</td>\n",
       "      <td>172</td>\n",
       "      <td>45</td>\n",
       "      <td>243</td>\n",
       "      <td>781</td>\n",
       "      <td>585</td>\n",
       "      <td>2068</td>\n",
       "      <td>1989</td>\n",
       "      <td>1449</td>\n",
       "      <td>...</td>\n",
       "      <td>2005</td>\n",
       "      <td>77</td>\n",
       "      <td>4212</td>\n",
       "      <td>12</td>\n",
       "      <td>181</td>\n",
       "      <td>8376</td>\n",
       "      <td>146</td>\n",
       "      <td>662</td>\n",
       "      <td>306</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Beta</th>\n",
       "      <td>3302</td>\n",
       "      <td>129</td>\n",
       "      <td>1093</td>\n",
       "      <td>7</td>\n",
       "      <td>50</td>\n",
       "      <td>455</td>\n",
       "      <td>3181</td>\n",
       "      <td>4324</td>\n",
       "      <td>2058</td>\n",
       "      <td>3914</td>\n",
       "      <td>...</td>\n",
       "      <td>5102</td>\n",
       "      <td>1855</td>\n",
       "      <td>11361</td>\n",
       "      <td>14</td>\n",
       "      <td>339</td>\n",
       "      <td>191</td>\n",
       "      <td>258</td>\n",
       "      <td>3953</td>\n",
       "      <td>787</td>\n",
       "      <td>649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Delta</th>\n",
       "      <td>746</td>\n",
       "      <td>605</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>7</td>\n",
       "      <td>189</td>\n",
       "      <td>440</td>\n",
       "      <td>509</td>\n",
       "      <td>769</td>\n",
       "      <td>379</td>\n",
       "      <td>...</td>\n",
       "      <td>691</td>\n",
       "      <td>98</td>\n",
       "      <td>2468</td>\n",
       "      <td>1</td>\n",
       "      <td>148</td>\n",
       "      <td>662</td>\n",
       "      <td>50</td>\n",
       "      <td>355</td>\n",
       "      <td>127</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ductal</th>\n",
       "      <td>6</td>\n",
       "      <td>70</td>\n",
       "      <td>14</td>\n",
       "      <td>54</td>\n",
       "      <td>23</td>\n",
       "      <td>11</td>\n",
       "      <td>29</td>\n",
       "      <td>131</td>\n",
       "      <td>68</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>43</td>\n",
       "      <td>11</td>\n",
       "      <td>73</td>\n",
       "      <td>96</td>\n",
       "      <td>3</td>\n",
       "      <td>115</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gamma</th>\n",
       "      <td>2</td>\n",
       "      <td>62</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>41</td>\n",
       "      <td>66</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>37</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>69</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Immune</th>\n",
       "      <td>8</td>\n",
       "      <td>22</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>104</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>...</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stellate</th>\n",
       "      <td>16</td>\n",
       "      <td>33</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>93</td>\n",
       "      <td>89</td>\n",
       "      <td>30</td>\n",
       "      <td>...</td>\n",
       "      <td>44</td>\n",
       "      <td>13</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>60</td>\n",
       "      <td>3</td>\n",
       "      <td>31</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 2025 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          RYR2    DMD  AC068491.1  PDGFD  LMO4  CLCN5  LYPLAL1  C12orf79  \\\n",
       "celltype                                                                   \n",
       "Acinar      98    239          30    322   136     73      348      1244   \n",
       "Alpha      125  10725         172     45   243    781      585      2068   \n",
       "Beta      3302    129        1093      7    50    455     3181      4324   \n",
       "Delta      746    605          33     64     7    189      440       509   \n",
       "Ductal       6     70          14     54    23     11       29       131   \n",
       "Gamma        2     62           1     10     4      9        6        41   \n",
       "Immune       8     22          47      3    12      5       20       104   \n",
       "Stellate    16     33           6      5    19     13       22        93   \n",
       "\n",
       "          MT-ND1  ZRANB2  ...  CLTC  RASGRF2   PCLO  CPA1  DEPTOR  MYO10  \\\n",
       "celltype                  ...                                              \n",
       "Acinar       159     246  ...   229        5    270   408      55    883   \n",
       "Alpha       1989    1449  ...  2005       77   4212    12     181   8376   \n",
       "Beta        2058    3914  ...  5102     1855  11361    14     339    191   \n",
       "Delta        769     379  ...   691       98   2468     1     148    662   \n",
       "Ductal        68      39  ...    43       11     73    96       3    115   \n",
       "Gamma         66      18  ...    37        9     82     0      20     37   \n",
       "Immune        54      40  ...    51        2     33     0       6     28   \n",
       "Stellate      89      30  ...    44       13     24     0       7     60   \n",
       "\n",
       "          TTC32   ME1  KCTD16  ETV1  \n",
       "celltype                             \n",
       "Acinar       18   126     132    14  \n",
       "Alpha       146   662     306   187  \n",
       "Beta        258  3953     787   649  \n",
       "Delta        50   355     127   221  \n",
       "Ductal        6    19       7     5  \n",
       "Gamma         6     9      69    20  \n",
       "Immune        3    23       1    12  \n",
       "Stellate      3    31       5     1  \n",
       "\n",
       "[8 rows x 2025 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "28f87e13-c036-481b-9038-84964660aa59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TF_Gene_tissue_Brain.csv',\n",
       " 'PO.txt',\n",
       " 'PO_brain.txt',\n",
       " 'readme.md',\n",
       " 'hg38.chrom.sizes',\n",
       " 'PP.txt',\n",
       " 'celltype_specific_cRE_interactions',\n",
       " 'TF_Gene_tissue_cutoff1.csv',\n",
       " 'all_tissue_SNP_Gene.txt',\n",
       " 'genes.protein.tss.tsv',\n",
       " 'trrust_rawdata.human.tsv',\n",
       " 'PP_brain.txt',\n",
       " 'model_init.pth']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scregat\n",
    "import os\n",
    "# This step adds tissue-specific Hi-C regulatory relationships.\n",
    "# The user needs to provide a set of files. For example,\n",
    "#in \n",
    "base_dir = '../data/'\n",
    "os.listdir(base_dir)\n",
    "# we include a file called PO_brain.txt,\n",
    "# which contains brain tissue-specific Hi-C links.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0c8efa11-da43-4ae3-838c-2c4095762055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6156 × 89233\n",
       "    obs: 'donor', 'celltype', 'celltype_rna'\n",
       "    var: '0', '1', '2', '3', '4', '5'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9f998ce1-4dea-4b29-a0dc-863b639716a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "only dataset_obj ...\n",
      "执行 liftover 命令: \"../liftOver\" \"../data/processed_files/peaks.sort.bed\" \"../hg19ToHg38.over.chain.gz\" \"../data/peaks_process/peaks_hg38.bed\" \"../data/peaks_process/unmap.bed\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      " ImplicitModificationWarning:/root/scReGAT/scregat/data_process.py:332: Trying to modify attribute `.var` of view, initializing view as actual.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing Hi-C ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " SettingWithCopyWarning:/root/scReGAT/scregat/data_process.py:607: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing TF ...\n",
      "additional TF...\n",
      "total candidate tf-gene:  28054\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/root/scReGAT/scregat/data_process.py:712: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:274.)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Note that regardless of whether the multi-omics data are matched at the single-cell level, \n",
    "# the regulatory graph will be constructed under the unmatched configuration. \n",
    "# Single-cell–level matched features will instead be incorporated during model training.\n",
    "\n",
    "\n",
    "dataset_obj = prepare_model_input(\n",
    "    # [Core Data] Single-cell ATAC-seq AnnData object.\n",
    "    # Requirement: .X must be Peak-by-Cell matrix; .obs must contain cell type annotations.\n",
    "    adata_atac = adata_atac,\n",
    "    \n",
    "    # [Output Path] Root directory for storing intermediate processed files.\n",
    "    # The script will create a 'processed_files' folder here (e.g., sorted bed files).\n",
    "    path_data_root = '../data/',\n",
    "    \n",
    "    # [File Reference] String path to the original ATAC file (used for naming/logging).\n",
    "    file_atac = ATAC_h5ad_file, \n",
    "    \n",
    "    # [Core Data] RNA expression matrix aggregated by cell type (Pseudo-bulk).\n",
    "    # Format: Pandas DataFrame. \n",
    "    # Index (Rows): Cell type names (must match 'celltype_rna' in adata_atac.obs).\n",
    "    # Columns: Gene Symbols (e.g., 'TP53').\n",
    "    df_rna_celltype = df_rna,\n",
    "    \n",
    "    # [Prior Knowledge] Path to eQTL (expression Quantitative Trait Loci) data.\n",
    "    # Used to link SNPs/non-coding regions to target genes.\n",
    "    path_eqtl = '../data/all_tissue_SNP_Gene.txt',\n",
    "    \n",
    "    # [Prior Knowledge] Suffix for Hi-C interaction files to specify tissue context.\n",
    "    # e.g., if set to \"_brain\", the code looks for 'PP_brain.txt' and 'PO_brain.txt'.\n",
    "    # PP = Promoter-Promoter, PO = Promoter-Other (Enhancer).\n",
    "    # Assuming that tissue-specific Hi-C data are not available, this option may be omitted. \n",
    "    Hi_C_file_suffix = \"\",  \n",
    "    # [Preprocessing] Whether to convert genomic coordinates from hg19 to hg38.\n",
    "    # Set True if input ATAC peaks are hg19 (requires LiftOver tool); False if already hg38.\n",
    "    hg19tohg38 = True,\n",
    "    liftover_path = '../liftOver',\n",
    "    chain_file = '../hg19ToHg38.over.chain.gz',\n",
    "    \n",
    "    # [QC Filter] Peak filtering threshold.\n",
    "    # Peaks must be accessible in at least 1% (0.01) of cells to be retained.\n",
    "    min_percent = 0.01,\n",
    "    \n",
    "    # [Prior Knowledge] Whether to use an extended TF (Transcription Factor) database.\n",
    "    # False: Uses TRRUST only (Curated, high-confidence).\n",
    "    # True: Uses TRRUST + CHEA3/ChIP-seq aggregated data.\n",
    "    use_additional_tf = True,\n",
    "    \n",
    "    # [Prior Knowledge] Reliability threshold for the extended TF database.\n",
    "    # Only applies if use_additional_tf=True.\n",
    "    # Integer indicating in how many tissues/datasets the TF-Gene link must appear to be kept.\n",
    "    # 10 indicates a high-confidence, conserved regulatory relationship.\n",
    "    tissue_cuttof = 10\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "148a168c-82b0-44ae-8e3f-a06a24f5d5de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Data(x=[29826, 1], edge_index=[2, 34203], y=[1], edge_tf=[359, 2], y_exp=[1634], cell='C0027_AAACAGCCACACTAAT-1')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_obj.list_graph[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3bfb0e25-d19a-455a-b7c6-45e5a3d5bf95",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_atac_test = '../data/dataset_atac_kRG_Pancreas.pkl'\n",
    "with open(file_atac_test, 'wb') as w_pkl:\n",
    "    str_pkl = pickle.dumps(dataset_obj)\n",
    "    w_pkl.write(str_pkl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb03058e-9092-4e07-ae64-4c8ed825de69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4cddf54d-8d88-4a5d-8ddd-0a32209b362b",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Training the Model and Obtaining Regulatory Scores\n",
    "\n",
    "You can train the model and generate regulatory scores directly from the command line.  \n",
    "For example, we used the following command to produce a sample dataset and a model weight file.  \n",
    "\n",
    "**Note:** In this example we explicitly enabled single‑cell–level matched features by using the `--use_sc_exp` flag together with the `--rna_file` argument:\n",
    "\n",
    "```bash\n",
    "python ./scReGAT/run_scregat_cli.py \\\n",
    "  --input_file ./dataset_atac_kRG_Pancreas.pkl \\\n",
    "  --output_file ./RS_score_Pancreas.h5ad \\\n",
    "  --use_sc_exp \\\n",
    "  --rna_file ./C0027_rna_celltype.h5ad \\\n",
    "  --save_model_path ./Pancreas.pth \\\n",
    "  --load_model_path ./scReGAT/data/model_init.pth \\\n",
    "  --gpu 2\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "## Argument Description\n",
    "\n",
    "```python\n",
    "def parse_args():\n",
    "    parser = argparse.ArgumentParser(description=\"Run scReGAT Model Training and Inference\")\n",
    "\n",
    "    # --- I/O parameters ---\n",
    "    parser.add_argument('--input_file', type=str, required=True, \n",
    "                        help='Path to the input ATAC pickle file (e.g., dataset_atac_core_MFG.pkl)')\n",
    "    parser.add_argument('--output_file', type=str, required=True, \n",
    "                        help='Path to save the output AnnData file (e.g., result.h5ad)')\n",
    "\n",
    "    # --- Single-cell expression integration parameters ---\n",
    "    parser.add_argument('--use_sc_exp', action='store_true',\n",
    "                        help='Enable integration of single-cell RNA expression data into graph node features')\n",
    "    parser.add_argument('--rna_file', type=str, default=None,\n",
    "                        help='Path to the RNA .h5ad file; required if --use_sc_exp is set')\n",
    "\n",
    "    # --- Model saving and loading ---\n",
    "    parser.add_argument('--save_model_path', type=str, default=None, \n",
    "                        help='Optional path to save trained model parameters')\n",
    "    parser.add_argument('--load_model_path', type=str, default=None, \n",
    "                        help='Optional path to load pre-trained model parameters')\n",
    "\n",
    "    # --- Training control ---\n",
    "    parser.add_argument('--skip_train', action='store_true',\n",
    "                        help='Skip the training phase and run inference directly')\n",
    "    parser.add_argument('--seed', type=int, default=1233, \n",
    "                        help='Random seed (default: 1233)')\n",
    "    parser.add_argument('--epochs', type=int, default=4, \n",
    "                        help='Number of training epochs (default: 4)')\n",
    "    parser.add_argument('--lr', type=float, default=1e-4, \n",
    "                        help='Learning rate (default: 1e-4)')\n",
    "    parser.add_argument('--batch_size', type=int, default=15, \n",
    "                        help='Training batch size (default: 15)')\n",
    "    parser.add_argument('--sparse_loss_weight', type=float, default=0.1, \n",
    "                        help='Weight for sparse loss (default: 0.1)')\n",
    "\n",
    "    # --- Testing / inference parameters ---\n",
    "    parser.add_argument('--test_batch_size', type=int, default=20, \n",
    "                        help='Batch size for inference (default: 20)')\n",
    "    parser.add_argument('--test_ratio', type=float, default=0.5, \n",
    "                        help='Ratio of cells to use for testing (default: 0.5)')\n",
    "\n",
    "    # --- Hardware parameters ---\n",
    "    parser.add_argument('--gpu', type=int, default=1, \n",
    "                        help='GPU ID to use; set -1 to use CPU (default: 1)')\n",
    "\n",
    "    return parser.parse_args()\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "## Recommendation\n",
    "\n",
    "We recommend users initialize training on new datasets using the model weight file:\n",
    "\n",
    "```\n",
    "./scReGAT/data/model_init.pth\n",
    "```\n",
    "\n",
    "This provides a stable starting point for training and helps ensure stability across experiments.  \n",
    "\n",
    "---\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scReGAT",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
